### ISSUE TITLE
Age key support

### ISSUE THREAD

**Issue by Tomaszal** [smlx/piv-agent#134]
As outlined in https://github.com/Mic92/ssh-to-age/issues/14, it would be great if it would be possible to use `piv-agent` to generate age keys in some way.

Currently the only way to get SSH keys from Yubikey seems to be [`age-plugin-yubikey`](https://github.com/str4d/age-plugin-yubikey). Unfortunately it doesn't seem to be able to use the PIV keys generated by `piv-agent`. While that itself isn't too big of a deal, it also seems that `age-plugin-yubikey` cannot run while `piv-agent` is active, as the PIV device is busy. That means that `piv-agent` would have to be stopped every time a file needs to be encrypted or decrypted, which is really inconvenient.

Depending on the outcome of https://github.com/Mic92/ssh-to-age/issues/14, it might be possible to get age keys from the SSH keys generated by `piv-agent`. However, I realised that `piv-agent` is already generating both SSH and GPG keys, so why not age? @str4d seems to acknowledge that [some sort of an agent](https://github.com/str4d/age-plugin-yubikey#agent-support) would be good to implement for age keys, and I have a hunch that `piv-agent` might already have a lot of the groundwork necessary for that.

**Comment by smlx** [comment]
Yes I'd like to implement this but the current specification for the plugin interface to `age` is still a work in progress https://github.com/C2SP/C2SP/pull/5

**Comment by Tomaszal** [comment]
Hi @smlx, it seems the specification has been merged last year, is this still something you would like to do? With `age` keys gaining support and `git` being able to use `ssh` keys for signing, it feels finally feasible to mostly avoid using `pgp` keys as a developer. It would be fantastic if `piv-agent` could provide `age` keys as imo it still has the best UX out of the physical security key [SSH solutions](https://github.com/FiloSottile/yubikey-agent?tab=readme-ov-file#alternatives) (including `yubikey-agent`), but the GPG experience isn't great (at not a fault of this project, GPG is just a pain to work with in general), which would be possible to avoid with `age`.

**Comment by smlx** [comment]
Thanks for the bump, and I totally agree that it would be amazing to eliminate GPG from common development workflows.

`age` support is still on my TODO list, but not a super high priority for now because ideally we should keep the x25519 key in hardware. Yubikeys are getting PIV support for x25519 [sometime soon](https://github.com/Yubico/yubico-piv-tool/blob/master/debian/changelog) so once I've got hardware with support I'll be able to implement this feature.

Yes technically `piv-agent` could use x25519 key files for this, but it just isn't very interesting because for me at least key files are only an emergency fallback mechanism.

**Comment by str4d** [comment]
Once I get access to a YubiKey with firmware 5.7.0+, I will add support for native age identities to `age-plugin-yubikey`, likely following the same pattern as I used for the P-256 slots (storing them in the reserved slots, making them identifiable in the self-signed certificate, and extracting PIN+touch policies from the YK cert extension). It would be great for `piv-agent` to be compatible with this!

It would also be good to know how `piv-agent`'s existing P-256 YubiKey support is incompatible with `age-plugin-yubikey`. How are you generating and managing those?

**Comment by smlx** [comment]
I agree it would be nice for `piv-agent` to be compatible with `age-plugin-yubikey`. I haven't looked into how you are doing PIN/touch policies, but in `piv-agent` it is purely convention as described in https://github.com/smlx/piv-agent/blob/main/internal/securitykey/slotspec.go

The generation of certificates in `piv-agent` is mostly cribbed from the way `yubikey-agent` does it.

Could you elaborate a bit on how you manage PIN/touch policies in cert extensions for `age-plugin-yubikey`?

**Comment by Tomaszal** [comment]
Yubico says "security keys with firmware 5.7 will be available to purchase in late May 2024" in their [latest press release](https://www.yubico.com/press-releases/yubicos-key-product-innovations-empower-enterprise-security-and-phishing-resistant-passwordless-authentication-at-scale/). They also confirmed Ed25519 and X25519 PIV support in the new firmware.

**Comment by Tomaszal** [comment]
Hello @smlx, just wanted to see if this is still on your radar?

**Comment by smlx** [comment]
Hi, yes it is. But the radar is full of other stuff too so finding the time to work on this is the problem :'(

**Comment by Tomaszal** [comment]
Totally understand that, didn't mean to rush! Just excited for this feature and wanted to check in

I wish I could help, but I'm quite busy myself and unfamiliar with Go. Perhaps I'll come around and try to  contribute at some point if it's not done by then, but can't promise anything
